### PR TITLE
BUG: fix broken error message in ujson.encode() (GH18878)

### LINKED ISSUE #18878: str vs bytes confusion in ujson on Python 3
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas._libs.json as ujson

ujson.encode(np.array(1))
```
#### Problem description

On CPython 3.5, this raises TypeError, as expected, but the output contains garbage bytes:
```
Traceback (most recent call last):
  File "bug.py", line 4, in <module>
    ujson.encode(np.array(1))
TypeError: � (0d array) is not JSON serializable at the moment
```
On pypy3, it segfaults.

The problem comes from this code: https://github.com/pandas-dev/pandas/blob/775099cde4a2424dfb369d79b8b6e2687e5af7f9/pandas/_libs/src/ujson/python/objToJSON.c#L1925-L1929 It calls `PyString_AS_STRING` (=`PyBytes_AS_STRING` on py3k) on the PyUnicodeObject returned by `PyObject_Repr`.

I haven't looked very carefully, but there seem to be a few more similar misuses of `PyString_AS_STRING` in the same file.

#### Expected Output

```
Traceback (most recent call last):
  File "bug.py", line 4, in <module>
    ujson.encode(np.array(1))
TypeError: array(1) (0d array) is not JSON serializable at the moment
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-135-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: fr_FR.UTF-8

pandas: 0.21.1
pytest: 3.2.5
pip: 9.0.1
setuptools: 37.0.0
Cython: None
numpy: 1.13.3
scipy: None
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by rlamy**
- [x] closes #18878
- [x] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/json/test_ujson.py [hunk 1]
```diff
@@ -780,8 +780,10 @@ def test_array_float(self):
         tm.assert_almost_equal(arr, arr_out)
 
     def test_0d_array(self):
-        with pytest.raises(TypeError):
+        msg = "array(1) (0d array) is not JSON serializable at the moment"
+        with pytest.raises(TypeError) as excinfo:
             ujson.encode(np.array(1))
+        assert str(excinfo.value) == msg
```
Comment (WillAyd) [inline pandas/tests/io/json/test_ujson.py]: Instead of doing this should be able to use the `match` argument of `pytest.raises`

**Review by WillAyd — CHANGES_REQUESTED** [review]
Nice! Can you add a whatsnew note for 1.0.0 as well?

File: pandas/tests/io/json/test_ujson.py (hunk 1 above)
Comment (rlamy) [reply pandas/tests/io/json/test_ujson.py]: I tried that first, actually, but the need to escape the regex makes the expected output harder to readamd the error report is less informative.

Comment (simonjayhawkins) [reply pandas/tests/io/json/test_ujson.py]: could use re.escape as we use this pattern elsewhere

```diff
diff --git a/pandas/tests/io/json/test_ujson.py b/pandas/tests/io/json/test_ujson.py
index 2cecf8f7b..d6572ac7b 100644
--- a/pandas/tests/io/json/test_ujson.py
+++ b/pandas/tests/io/json/test_ujson.py
@@ -780,10 +780,10 @@ class TestNumpyJSONTests:
         tm.assert_almost_equal(arr, arr_out)
 
     def test_0d_array(self):
-        msg = "array(1) (0d array) is not JSON serializable at the moment"
-        with pytest.raises(TypeError) as excinfo:
+        # gh-18878
+        msg = re.escape("array(1) (0d array) is not JSON serializable at the moment")
+        with pytest.raises(TypeError, match=msg):
             ujson.encode(np.array(1))
-        assert str(excinfo.value) == msg
 
     @pytest.mark.parametrize(
         "bad_input,exc_type,kwargs",

```

File: pandas/tests/io/json/test_ujson.py [hunk 2]
```diff
@@ -780,8 +780,10 @@ def test_array_float(self):
         tm.assert_almost_equal(arr, arr_out)
 
     def test_0d_array(self):
```
Comment (simonjayhawkins) [inline pandas/tests/io/json/test_ujson.py]: can you add the issue number.

**Review by simonjayhawkins — COMMENTED** [review]
@rlamy Thanks for the PR. lgtm. couple of suggestions.

File: pandas/tests/io/json/test_ujson.py (hunk 2 above)
Comment (rlamy) [reply pandas/tests/io/json/test_ujson.py]: Done. Thanks for reviewing.

**Review by simonjayhawkins — APPROVED** [review]
lgtm @WillAyd

**Comment by jreback** [comment]
thanks @rlamy

**Pull request merged**